### ISSUE TITLE
feature request: support swc

### ISSUE THREAD

**Issue by punkeel** [fkling/astexplorer#555]
> swc is a super-fast compiler written in rust; producing widely-supported javascript from modern standards and typescript.

swc is a serious project with 10k stars on [GitHub](https://github.com/swc-project/swc). It is used in [Deno](https://github.com/denoland/deno) to transpile TypeScript code into JS. I'm not 100% sure, but there may be a WebAssembly version exposing parse / codegen functions.

It would be great to add support for swc in astexplorer.

cc @kdy1

**Comment by RReverser** [comment]
> I'm not 100% sure, but there may be a WebAssembly version exposing parse / codegen functions.

It seems to have it right in the tree you provided: https://github.com/swc-project/swc/tree/master/wasm. If you can find or publish an npm package for bindings (see my https://github.com/RReverser/astexplorer-syn for an example), it should be fairly straightforward to integrate into ASTExplorer.

**Comment by punkeel** [comment]
Indeed, `swc/wasm` exists and is deployed to npm. It's targeting nodejs, so I've opened https://github.com/swc-project/swc/issues/1277 to add a web version of it. I'll try playing a bit with AST Explorer to see if I can add support there, thanks for the syn pointer!

**Comment by punkeel** [comment]
I've pushed a WIP pr #556 that has the following known issues:

1. It doesn't build, as `@swc/wasm` isn't meant to be used from browsers. To test it locally, I've manually built a browser-friendly version and replaced the `node_modules/...` files. See issue linked above in swc project.
2. When the code (or a setting) is changed, Autofocus is broken. Would you have any ideas what could cause this?

Thinking it could be WebAssembly shenanigans, I've tried wrapping the `parseSync` call in `JSON.parse(JSON.stringify(...))`, but it didn't help.

Aside from that, it was really easy to integrate in AST Explorer, ✨  magic ✨ !
<img width="1663" alt="swc astexplorer" src="https://user-images.githubusercontent.com/1859135/102410630-18c27f00-3ff1-11eb-8a69-bd59793b1af6.png">

**Comment by RReverser** [comment]
> Aside from that, it was really easy to integrate in AST Explorer, ✨ magic ✨ !

Glad you liked it :) I've set it up a long time ago, and sometimes felt bad about all the Webpack magic it's doing, but it does seem to make contributions easier.

**Comment by punkeel** [comment]
A first version of `@swc/wasm-web` was published, but another release is required to fix a bug, see https://github.com/swc-project/swc/issues/1291.

I've started working on a swc transformer, and it was just as easy as adding a parser to AST Explorer (yay!). :)
![image](https://user-images.githubusercontent.com/1859135/103159525-cd675800-47ca-11eb-8297-a35e01caa386.png)

I'll update the PR to include this transformer, and I've prepared another PR (https://github.com/fkling/astexplorer/pull/557) to replace `http-server` in dev with `light-server`, the former sending the proper Accept header for `.wasm` files.

cc @9oelM

**Comment by 9oelM** [comment]
@punkeel hi!

It seems that swc does not run something like a nightly build, which is the reason I had to build web wasm of swc myself and manually include it in 9oelm.github.io/swc-wasm-demo. It'd be great if swc would support nightly build (I'm open to helping if possible, too), although I'm not sure if this is the proper repo to talk about this 😅

Also i find that we got a lot to work on the wasm side of swc like what you said

**Comment by ShayRubach** [comment]
@9oelM TLDR - Is the swc type for ASTExplorer is fully working? And if so, how do I get an access to it? Doesn't seem like it is officially published (can't see 'swc' type there).

**Comment by 0xdevalias** [comment]
Given that the following PR has been merged and `swc` seems to be supported now, should this issue be closed?

- https://github.com/fkling/astexplorer/pull/637

![image](https://github.com/fkling/astexplorer/assets/753891/6615f770-1c52-4a1b-b2a1-d51a01c8ffcc)

**Comment by mantou132** [comment]
Currently "Autofocus" does not work when `decorators` are enabled?
